### PR TITLE
Add testing boilerplate, do not attempt to resolve abstract models by default

### LINKED ISSUE #41: model:typer command tries to write abstract models
Running the `php artisan model:typer` command results in the following error if your project has an abstract model class:

```php
   TypeError 

  FumeApp\ModelTyper\Actions\RunModelShowCommand::__invoke(): Return value must be of type array, null returned

  at vendor/fumeapp/modeltyper/src/Actions/RunModelShowCommand.php:30
     26▕         if ($exitCode !== 0) {
     27▕             throw new Exception('You may need to install the doctrine/dbal package to use this command.');
     28▕         }
     29▕ 
  ➜  30▕         return json_decode(Artisan::output(), true);
     31▕     }
     32▕ }
     33▕

      [2m+20 vendor frames [22m
  21  artisan:35
      Illuminate\Foundation\Console\Kernel::handle()
```

I ran into this issue when I added an abstract class called `BaseModel` that extended the eloquent model.

Suggestions:
1) Do not attempt to write abstract classes
2) Create a validation check for the `show:model` command output in `RunShowModelCommand` action to avoid confusing error messages when artisan fails. 

Point number 2 is also relevant for the future testing, since `Artisan::output()` returns empty string in Laravel test environment unless you have set the following property in your test case class.
```php
public $mockConsoleOutput = false;
```

We could do a validity check along the lines of 
```php
$output = json_decode(Artisan::output(), true);

if (empty($output)) {
	$msg = "Could not resolve types for model '$model', Artisan::output() is empty.";
	$msg .= PHP_EOL . 'If you are running tests, make sure to set {public $mockConsoleOutput = false;}';
	throw new \Exception($msg);
}
```

### PR DESCRIPTION
**Pull request by nonetallt**
### Add proper testing setup for the package. 

Impact of the changes to the end-user should be minimal. Only change to the command is the addition of `--resolve-abstract` option that is now longer enabled by default. 

I decided against simply filtering abstract models since the `ShowModel` command attempts to resolve models from the application container, which means that it's possible for the user to have an abstract model bound to an instance in their container.

### Changes:

- Add testing setup
- Add laravel-skeleton for testing
- Add test watcher
- Fix requirements and prune unused testing packages
- Do not attempt to resolve abstract models by default (fixed #41) 
- Extract `Generator::getModels()` into a separate action `GetModels`
- Improved model file name resolution supports fully qualified class names.
- `GetModels` has an option to include or exclude models, but the option is currently unused.
- Properly handle errors from the `ShowModelCommand`

Run tests with `composer test` or `composer test-watch`. I've added a couple of basic test as examples and to make sure that things work in general. Hopefully someone else can write more throughout coverage for the actual command output. Feel free to setup CI testing.

### Problems:

Unfortunately the way Laravel handles output from nested commands makes debugging command output in the tests very obnoxious because `Artisan::output()` does not work after it's been called inside a command. This happens regardless of whether or not `$mockConsoleOutput` is being used. Note that this does not affect the actual output of the command in a live environment nor does it affect the output expectations of tests. You simply can't see the command output by calling `Artisan::output()` in the test after calling the `model:typer` command.

The alternative would be to pass the command to the underlying action and then use `$command->call()` instead of `Artisan::call()` but that doesn't work for the `ShowModel` command because we need the output immediately (and calling another command using a command routes the output into the output of that command).

### Suggestions:

It might be a good idea to add a new `--output-file` option that will direct the output of the command to a specified file rather than printing directly to console output. This can then be used for easier test debugging as well as general use for the end-user. I already added a trait that can clean the output between tests.

### CODE REVIEW

**Comment by tcampbPPU** [comment]
Thanks for the PR @nonetallt this was much needed! Let me take a look over things and provide any feedback if needed.

**Review by tcampbPPU — APPROVED** [review]
There are a few composer stan issues when running `composer stan` but these look pretty minor and could be resolved in other PR @acidjazz so we dont have to hold this 1 up. Also in another PR I will setup GH actions to automate testing like we do in Humble.

For the rest looks good! Thanks again @nonetallt for helping us out!

**Comment by acidjazz** [comment]
@nonetallt can you by chance throw an action in here that kicks these tests off? 

Here is an example of one https://github.com/fumeapp/laranuxt/blob/main/.github/workflows/test-php.yml

you can kill the branches/paths part etc

**Comment by acidjazz** [comment]
or better yet https://github.com/fumeapp/humble/blob/master/.github/workflows/test.yml

File: src/Actions/Generator.php [hunk 1]
```diff
@@ -2,9 +2,8 @@
 
 namespace FumeApp\ModelTyper\Actions;
 
+use FumeApp\ModelTyper\Exceptions\ModelTyperException;
 use Illuminate\Support\Collection;
-use Illuminate\Support\Facades\File;
-use Symfony\Component\Finder\SplFileInfo;
```
Comment (tcampbPPU) [inline src/Actions/Generator.php:7]: Need to add back this import to fix PHP stan error

File: src/Actions/RunModelShowCommand.php [hunk 2]
```diff
@@ -2,31 +2,113 @@
...
+     *
+     * @return bool $mocked
+     */
+    private function outputMocked(): bool
+    {
+        if (! app()->runningInConsole()) {
+            return false;
+        }
+
+        if (! config('app.env') === 'testing') {
```
Comment (tcampbPPU) [inline src/Actions/RunModelShowCommand.php:108]: maybe you can replace with
```php
if (! app()->environment('testing')) {
 // ...
}
```

**Comment by tcampbPPU** [comment]
@nonetallt  dont worry about the GH action stuff I will take care of it, there was a few extra I wanted to add

**Comment by tcampbPPU** [comment]
@nonetallt alright there are 2 composer stan issues left, pretty sure [this suggestion](https://github.com/fumeapp/modeltyper/pull/42#discussion_r1151275775) will fix that.
For the other one I am not sure, but I think we can ignore.

Thanks again for helping us out :)

File: src/Actions/RunModelShowCommand.php (hunk 2 above)
Comment (nonetallt) [reply src/Actions/RunModelShowCommand.php]: Resolved by [18525a3](https://github.com/fumeapp/modeltyper/pull/42/commits/18525a3ead34a9fe1b845b7afaf96993398790dc)

**Comment by nonetallt** [comment]
I removed the following composer hook script:

```json
{
  "scripts": {
    "post-autoload-dump": [
      "@php vendor/bin/testbench package:discover --ansi"
    ]
  }
}
```

Running the command is not required for testing because of the copied laravel-skeleton in `test/laravel-skeleton`. The command creates a symlink between the package vendor and testbench that can be problematic for local testing. See https://github.com/laravel/vite-plugin/issues/210 for more details.
